### DISCUSSION TITLE
Multiple schemas

### DISCUSSION THREAD

**Discussion by timothyarmes** [0no-co/GraphQLSP#223] [Q&A] [answered]
Is there a way to declare multiple graphQL endpoints?

**Comment by JoviDeCroock** [comment]
No, what is your use case for this? My current philosophy for GraphQL is that it already functions as a way to combine endpoints into one common facade for your front-end so having multiple endpoints is a use-case I haven't actually seen before. It would be helpful if you elaborated a bit more on the why

**Reply by timothyarmes** [reply]
A simple use case is for a client fetching GraphQL data from two completely unrelated servers.  A weather app fetching weather information from one public API, and map data from another, for instance.

In my particular case we have a production server for most queries, and a dev server for unpublished in-dev endpoints.

**Comment by Yuripetusko** [comment]
Encountered the same issue now. Have 2 unrelated graphql endpoints that I don't have a control over and so have 2 schemas, but having issues setthings this up so far. 

The project builds fine, and if I ctrl+click into the column, it jumps to the right part of correct schema, but intellij still shows it as error and as if it still sees only 1 of the schemas globally

<img width="661" alt="image" src="https://github.com/0no-co/GraphQLSP/assets/666169/dd552fb5-46bb-4bb9-a47f-b487fc5d7a1b">

**Reply by Yuripetusko** [reply]
actually weirdly it works fine, just that typescript in IDE complains

**Reply by ivari** [reply]
Doesn't everything inside the graphql method depend on the LSP so currently IDE is the only place it is even supposed to work? :)

**Reply by Yuripetusko** [reply]
right, it only suppose to work in IDE you are right. However the issue still stands that it doesn't understand multuple of the same plugins defined in tsconfig with different schemas

**Comment by ivari** [comment]
I'd love to jump on this discussion, with more details about trying to custom-implement the multiple schemas on `gql.tada` discussions – https://github.com/0no-co/gql.tada/discussions/178

Implementing multiple external GQL APIs is one thing, but in our case we have a separate end-user and admin UI endpoints. Not going to argue that there may be better ways to do the separation than that.

**Reply by kitten** [reply]
Hiya 👋 Just here to acknowledge that, yes, this will need to happen and be implemented, and we have a good idea on how to make this happen. I haven't written up an RFC for this yet, but that's because this is likely something we're going to get to after these two issues, which also both require LSP+CLI+tada coordination:
- https://github.com/0no-co/gql.tada/issues/77
- https://github.com/0no-co/gql.tada/issues/160

After these two issues are done, we'll be in a good state to revisit this, and will likely implement this by passing through `schema` information to tada's introspection output, which then the LSP (and CLI) can in turn pick up again via the document’s types, to check which schema the documents belong to.

**Reply by HiggsWRX** [reply]
Hey @kitten, it's great to see such good progress on this library. 

In the mean time, do you have any suggestions on how we could circumvent this? A way to merge our json schemas and pass just the one to gql.tada, or something else?

**Reply by kitten** [reply]
You can do the following:
- make sure you use `initGraphQLTada<>()` to create a `graphql` function for each schema you have
- give each function a unique name (set `"template": "yourFnName"`) in the `tsconfig.json`
- then use that name to import your custom `graphql` function
- repeat the plugin in `plugins` in your `tsconfig.json` for each schema

This **will** not be supported by the CLI and lead to some issues, but that's the compromise you'll have to live with until it's supported in the plugin itself

**Reply by tianyingchun** [reply]
It is now possible to define two independently schemes?

**Reply by kitten** [reply]
No, that's still a workaround. What I wrote above is not a supported use-case and the original comment still stands. This is a feature we'll implement and until then the `template` steps are just to get some things working — that's not reflective of what multi-schema support would look like, and it will lead to some issues, especially in the CLI

**Reply by tianyingchun** [reply]
This feature is very useful. Do we have a specific development plan? (^-^) 👍

**Comment by kitten** [comment, marked as answer]
Multi-Schema Support has been released in `gql.tada` 1.6.0 (+ `@0no-co/graphqlsp@^0.12.0`).
There's more information in the corresponding announcement devlog post: https://gql-tada.0no.co/devlog/2024-04-26

**Reply by ViliamKopecky** [reply]
Thank you so much. Do you have any public working examples? I prepared my own test which partially works, but it doesn't quite do everything I expected... maybe I'm doing something wrong?

https://github.com/ViliamKopecky/gqltada-multi-schema-test

**Reply by kitten** [reply]
@ViliamKopecky: Just issued a patch. Basically, namespace imports were not recognized correctly and we didn't add a default compiler options (specifically, `moduleResolution`) which stopped the CLI from working correctly. Once you upgrade the packages, it should all work correctly 

By the way, I spotted some comments along the lines of that `tsc` reports no errors:
> neither VSCode or `npm run typecheck` reports an error here but they should

Diagnostics from GraphQLSP will never show up in `tsc` / TypeScript output. Language server plugins cannot manipulate the TypeScript compiler output and GraphQLSP only provides a better in-editor DX. If you want to check for `gql.tada` errors outside of your editor, you'll have to additionally run `gql.tada check`, if type checks aren't sufficien for you.

Apart from that, upgrading seems to fix everything in your repo ✨

**Reply by ViliamKopecky** [reply]
@kitten wow, thank you so much!

**Comment by Yuripetusko** [comment]
Amazing! great work team, thanks very much for that

**Comment by ivari** [comment]
I heard somewhere it took you three years to finalise the implementation for this library to work? … and now you are just cranking out DX improvements weekly. Amazing work. Thank you so much!

Can't wait to try this out.

_I'll try to finally figure out the paperwork I need for my company to sponsor this thing._

**Reply by kitten** [reply]
Thanks for the high praise ❤️ 
I think we did first discuss `gql.tada`+`graphqlsp` (or rather the DX that we now have) about three years ago, and were then slacking a little and didn't move on implementing it. It did take us quite a while, but we're super excited now to be able to iterate on this and knock out the last few problems 🙌

**Comment by tianyingchun** [comment]
works fine :) nice feature ❤️

**Comment by ivari** [comment]
Can confirm it works great.
_Had to reload VS Code once for it to pick up proper typings after schema switch, but that's #justIDEthings_
